### DISCUSSION TITLE
how to deal with instance with 3 points in MMOCR 1.x det

### DISCUSSION THREAD

**Discussion by jorie-peng** [open-mmlab/mmocr#1595] [Q&A]
dataset:  total-text
env: MMOCR dev-1.x
model: fcenet
config: as same as fcenet_fpn_ic15
error: because total-text has instance with 1/2/3 points,

step1. if train fcenet with dev-1.x totaltext dataset directly, it will have error , it seems must have points > 6。

> TopologyException: Input geom 0 is invalid: Self-intersection at 95.599836545279544 343.73606162303474
> Traceback (most recent call last):
>   File "tools/train.py", line 117, in <module>
>     main()
>   File "tools/train.py", line 113, in main
>     runner.train()
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/runner/runner.py", line 1631, in train
>     model = self.train_loop.run()  # type: ignore
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/runner/loops.py", line 88, in run
>     self.run_epoch()
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/runner/loops.py", line 103, in run_epoch
>     for idx, data_batch in enumerate(self.dataloader):
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 435, in __next__
>     data = self._next_data()
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1065, in _next_data
>     return self._process_data(data)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1111, in _process_data
>     data.reraise()
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/_utils.py", line 428, in reraise
>     raise self.exc_type(msg)
> AssertionError: Caught AssertionError in DataLoader worker process 1.
> Original Traceback (most recent call last):
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 198, in _worker_loop
>     data = fetcher.fetch(index)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
>     data = [self.dataset[idx] for idx in possibly_batched_index]
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
>     data = [self.dataset[idx] for idx in possibly_batched_index]
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/dataset/base_dataset.py", line 408, in __getitem__
>     data = self.prepare_data(idx)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/dataset/base_dataset.py", line 789, in prepare_data
>     return self.pipeline(data_info)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/dataset/base_dataset.py", line 58, in __call__
>     data = t(data)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmcv/transforms/base.py", line 12, in __call__
>     return self.transform(results)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmcv/transforms/processing.py", line 1551, in transform
>     results = self.resize(results)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmcv/transforms/base.py", line 12, in __call__
>     return self.transform(results)
>   File "/mnt/lustre/pengjiahui/download_dataset/OCR/mmocr/mmocr/datasets/transforms/ocr_transforms.py", line 604, in transform
>     self._resize_polygons(results)
>   File "/mnt/lustre/pengjiahui/download_dataset/OCR/mmocr/mmocr/datasets/transforms/ocr_transforms.py", line 584, in _resize_polygons
>     polygon = crop_polygon(polygon, crop_bbox)
>   File "/mnt/lustre/pengjiahui/download_dataset/OCR/mmocr/mmocr/utils/polygon_utils.py", line 153, in crop_polygon
>     poly = poly2shapely(polygon)
>   File "/mnt/lustre/pengjiahui/download_dataset/OCR/mmocr/mmocr/utils/polygon_utils.py", line 111, in poly2shapely
>     assert polygon.size % 2 == 0 and polygon.size >= 6
> AssertionError

step2:  refer to https://github.com/open-mmlab/mmocr/issues/768, change instance with points<8 to ignore=True, add filter code in /mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/dataset/base_dataset.py `filter_data()`, but still fail. 

> TopologyException: Input geom 0 is invalid: Self-intersection at 143.39978858616797 515.82527798853891
> Traceback (most recent call last):
>   File "tools/train.py", line 117, in <module>
>     main()
>   File "tools/train.py", line 113, in main
>     runner.train()
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/runner/runner.py", line 1631, in train
>     model = self.train_loop.run()  # type: ignore
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/runner/loops.py", line 88, in run
>     self.run_epoch()
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/runner/loops.py", line 103, in run_epoch
>     for idx, data_batch in enumerate(self.dataloader):
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 435, in __next__
>     data = self._next_data()
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1065, in _next_data
>     return self._process_data(data)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1111, in _process_data
>     data.reraise()
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/_utils.py", line 428, in reraise
>     raise self.exc_type(msg)
> shapely.errors.TopologicalError: Caught TopologicalError in DataLoader worker process 3.
> Original Traceback (most recent call last):
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 198, in _worker_loop
>     data = fetcher.fetch(index)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
>     data = [self.dataset[idx] for idx in possibly_batched_index]
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
>     data = [self.dataset[idx] for idx in possibly_batched_index]
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/dataset/base_dataset.py", line 413, in __getitem__
>     data = self.prepare_data(idx)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/dataset/base_dataset.py", line 794, in prepare_data
>     return self.pipeline(data_info)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmengine/dataset/base_dataset.py", line 58, in __call__
>     data = t(data)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmcv/transforms/base.py", line 12, in __call__
>     return self.transform(results)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmcv/transforms/processing.py", line 1551, in transform
>     results = self.resize(results)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/mmcv/transforms/base.py", line 12, in __call__
>     return self.transform(results)
>   File "/mnt/lustre/pengjiahui/download_dataset/OCR/mmocr/mmocr/datasets/transforms/ocr_transforms.py", line 604, in transform
>     self._resize_polygons(results)
>   File "/mnt/lustre/pengjiahui/download_dataset/OCR/mmocr/mmocr/datasets/transforms/ocr_transforms.py", line 584, in _resize_polygons
>     polygon = crop_polygon(polygon, crop_bbox)
>   File "/mnt/lustre/pengjiahui/download_dataset/OCR/mmocr/mmocr/utils/polygon_utils.py", line 155, in crop_polygon
>     poly_cropped = poly.intersection(crop_poly)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/shapely/geometry/base.py", line 695, in intersection
>     return geom_factory(self.impl['intersection'](self, other))
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/shapely/topology.py", line 73, in __call__
>     self._check_topology(err, this, other)
>   File "/mnt/cache/pengjiahui/anaconda3/envs/mmde3/lib/python3.8/site-packages/shapely/topology.py", line 38, in _check_topology
>     raise TopologicalError(
> shapely.errors.TopologicalError: The operation 'GEOSIntersection_r' could not be performed. Likely cause is invalidity of the geometry <shapely.geometry.polygon.Polygon object at 0x7f13b8ee6700>
 
step3:  for TopologicalError error, refer to https://github.com/open-mmlab/mmocr/issues/1388, change pipeline, but fail still, error message keep same as step2.

so how to change my config or dataset, can train successfully?

**Comment by Harold-lkk** [comment]
Thanks for your report! Can you share the code you change the `filter_data`?

**Reply by jorie-peng** [reply]
`filter_data` code as follows:

```
def filter_data(self) -> List[dict]:
    for data_info in self.data_list:
        for j,instance_info in enumerate(data_info['instances']):
             polygon = instance_info['polygon']
             if len(polygon)<8:
                 del data_info['instances'][j]
    return self.data_list
```

**Reply by Harold-lkk** [reply]
Thanks for your reply! We will try it and reply ASAP.

**Reply by gaotongxiao** [reply]
We have thoroughly fixed this issue, and also released an example config for totaltext:

https://github.com/open-mmlab/mmocr/blob/dev-1.x/configs/textdet/fcenet/fcenet_resnet50_fpn_1500e_totaltext.py

**Reply by jorie-peng** [reply]
Thanks! I will try it!
